django/db/models/constraints.py

```diff
@@ -666,6 +664,10 @@ def validate(self, model, instance, exclude=None, using=DEFAULT_DB_ALIAS):
                             raise ValidationError(
                                 instance.unique_error_message(model, self.fields),
                             )
+                raise ValidationError(
+                    self.get_violation_error_message(),
+                    code=self.violation_error_code,
+                )
```

Comment:
This change was necessary to allow non-conditional unique constraints over fields validation to be tested without actually creating the constraint. The fact that no `ValidationError` could be raised under some conditions was definitely overlooked.

That likely explains why the initial `nulls_distinct` tests were added to `tests/validation` instead of `tests/constraints` in the first place https://github.com/django/django/pull/17058/files#diff-adae729932de755cd1b0e4a4413203256a3b17864523cba7ea646ae6d9951c0bR222

Reply:
I was wondering why the tests were spread across `tests/validation` and `tests/constraints`. Once all these `UniqueConstraint` tickets are resolved, is there a need to move most of the tests to `tests/constraints`?

Reply:
The `if not self.expressions:` clause looks a bit too clever. Is it reasonable to assume that the `model` passed in to `UniqueConstraint.validate(model, ...)` is the model class which is relevant for this constraint instance? If we refactor the code to remove the call to  `instance.get_constraints()` as below then no tests fail in the suite (other than a difference of error message in  `test_validate_fields_unattached`).

From what I can see `constraint.validate()` is always called with the correct `model` class by `Model.validate_constraints()`. Would the below not be clearer?
```python
...
if queryset.exists():
    if self.fields:
        raise ValidationError(
            instance.unique_error_message(model, self.fields),
        )
    else:
        raise ValidationError(
            self.get_violation_error_message(),
            code=self.violation_error_code,
        )
```
I'm happy to raise a separate PR if you think this one isn't the place... or happy to be taught about the edge case which I'm missing :-)